Conditional login through identity provider

dmartinol opened the pull request:
Closes #20191

Added new fields in the OIDC settings page to specify an additional authentication check based on token claims:
- one switch to toggle the check on and off: defaults to off
- one field to define the token name (must be in the ID token section): defaults to empty string
- one field to define the expected token value (with regex matching): defaults to empty string
Latest two fields are hidden if the switch is turned off.

Default values are provided in the IdentityProviderModel, which should guarantee proper upgrade of data from previous release.

I have found no existing unit test on this area, so I did not include a new one for this feature. 

Updated Server Admin guide to reflect the change.

Tested on local environment and also on OpenShift deployment.

dmartinol committed:
This is to close [KEYCLOAK-19956](https://issues.redhat.com/browse/KEYCLOAK-19956): Conditional login through identity provider

Added new fields in the OIDC settings page to specify an additional authentication check based on token claims:
- one switch to toggle the check on and off: defaults to off
- one field to define the token name (must be in the ID token section): defaults to empty string
- one field to define the expected token value (with regex matching): defaults to empty string
Latest two fields are hidden if the switch is turned off.

Default values are provided in the IdentityProviderModel, which should guarantee proper upgrade of data from previous release.

I have found no existing unit test on this area, so I did not include a new one for this feature.

Updated Server Admin guide to reflect the change.

Tested on local environment and also on OpenShift deployment.

dmartinol commented:
> This is to close [KEYCLOAK-19956](https://issues.redhat.com/browse/KEYCLOAK-19956)

Created #20191

edewit commented on js/apps/admin-ui/src/identity-providers/add/AdvancedSettings.tsx:
```diff
@@ -93,8 +98,14 @@ type AdvancedSettingsProps = { isOIDC: boolean; isSAML: boolean };
+    formState: { errors },
+  } = useFormContext<IdentityProviderRepresentation>();
   const [syncModeOpen, setSyncModeOpen] = useState(false);
+  const [filteredByClaim, setFilteredByClaim] = useState(control._formValues["config"]["filteredByClaim"]);
```
Don't use "private" fields. You should replace this by a "watch"
```suggestion
  const filteredByClaim = useWatch({ control, name: "config.filteredByClaim", defaultValue: false});
```

edewit commented on js/apps/admin-ui/src/identity-providers/add/AdvancedSettings.tsx:
```diff
@@ -125,6 +136,87 @@ export const AdvancedSettings = ({ isOIDC, isSAML }: AdvancedSettingsProps) => {
+                  field.onChange(v)
+                  setFilteredByClaim(v)
+                  setClaimFilterRequired(value)
+                  console.debug('setFilteredByClaim ' + v)
```
remove debug

dmartinol replied:
thanks

edewit commented on js/apps/admin-ui/src/identity-providers/add/AdvancedSettings.tsx:
```diff
@@ -125,6 +136,87 @@ export const AdvancedSettings = ({ isOIDC, isSAML }: AdvancedSettingsProps) => {
+              type="text"
+              id="kc-claim-filter-name"
+              data-testid="claimFilterName"
+              {...register("config.claimFilterName", { required: claimFilterRequired })}
```
```suggestion
              {...register("config.claimFilterName", { required: true })}
```

edewit commented on js/apps/admin-ui/src/identity-providers/add/AdvancedSettings.tsx:
```diff
@@ -125,6 +136,87 @@ export const AdvancedSettings = ({ isOIDC, isSAML }: AdvancedSettingsProps) => {
+            helperTextInvalid={t("common:required")}
+          >
+            <KeycloakTextInput
+              isRequired={claimFilterRequired}
```
because `claimFilterRequired` is always true this can be shorted:
```suggestion
              isRequired
```

dmartinol replied:
thanks

edewit commented on js/apps/admin-ui/src/identity-providers/add/AdvancedSettings.tsx:
```diff
@@ -125,6 +136,87 @@ export const AdvancedSettings = ({ isOIDC, isSAML }: AdvancedSettingsProps) => {
+            }
+            helperTextInvalid={t("common:required")}
+          >
+            <KeycloakTextInput
```
also add the validated property here so the field turns red when there is an error

edewit commented on js/apps/admin-ui/src/identity-providers/add/AdvancedSettings.tsx:
```diff
@@ -125,6 +136,87 @@ export const AdvancedSettings = ({ isOIDC, isSAML }: AdvancedSettingsProps) => {
+          >
+            <KeycloakTextInput
+              isRequired={claimFilterRequired}
+              isReadOnly={!claimFilterRequired}
```
no need to add this as `false` is the default and `claimFilterRequired` is always true

edewit commented on js/apps/admin-ui/src/identity-providers/add/AdvancedSettings.tsx:
```diff
@@ -125,6 +136,87 @@ export const AdvancedSettings = ({ isOIDC, isSAML }: AdvancedSettingsProps) => {
+            <KeycloakTextInput
+              isRequired={claimFilterRequired}
+              isReadOnly={!claimFilterRequired}
+              type="text"
```
same here `type="text"` is the default no need to add it

edewit commented on js/apps/admin-ui/src/identity-providers/add/AdvancedSettings.tsx:
```diff
@@ -125,6 +136,87 @@ export const AdvancedSettings = ({ isOIDC, isSAML }: AdvancedSettingsProps) => {
+              {...register("config.claimFilterName", { required: claimFilterRequired })}
+            />
+          </FormGroup>
+          <FormGroup
+            label={t("identity-providers:claimFilterValue")}
+            labelIcon={
+              <HelpItem
+                helpText={t("identity-providers-help:claimFilterValue")}
+                fieldLabelId="identity-providers:claimFilterName"
+              />
+            }
+            fieldId="kc-claim-filter-value"
+            isRequired={claimFilterRequired}
+            validated={
+              errors.config?. claimFilterValue
+                ? ValidatedOptions.error
+                : ValidatedOptions.default
+            }
+            helperTextInvalid={t("common:required")}
+          >
+            <KeycloakTextInput
+              isRequired={claimFilterRequired}
+              isReadOnly={!claimFilterRequired}
+              type="text"
+              id="kc-claim-filter-value"
+              data-testid="claimFilterValue"
+              {...register("config.claimFilterValue", { required: claimFilterRequired })}
+            />
+          </FormGroup>
```
this field has similar problems

edewit reviewed (changes requested):
I only looked at the frontend code, don't know about this feature

dmartinol committed:
integrated suggested changes

edewit reviewed (dismissed):
:+1:

dmartinol committed:
Merge branch 'main' into main

stianst commented:
@pedroigor @mposolda could you take a look at this please?

stianst commented:
Had a quick look at this and couple notes from me:

* General approach looks sane to me, but not sure about the names/descriptions of the fields. Filter by claim is not directly clear to me that it means require claim to be present. Perhaps it could be renamed to something like `Require claim to authenticate` and `Required claim name` / `Required claim value`?
* We need some tests for this, not sure how we can do that, but I imagine we could extend on testing with IdP brokering with another realm as that would be easier to configure/setup

mposolda commented on js/apps/admin-ui/src/identity-providers/add/AdvancedSettings.tsx:
```diff
@@ -125,6 +136,90 @@ export const AdvancedSettings = ({ isOIDC, isSAML }: AdvancedSettingsProps) => {
       />
       <SwitchField field="config.hideOnLoginPage" label="hideOnLoginPage" />
 
+      <FormGroupField label="filteredByClaim">
```
The feature is implemented just for OIDC Identity providers. So are we showing the relevant options just for OIDC Identity providers? As I think it should be hidden for SAML IDP and also for Social Identity providers, which are not OIDC Identity provider.

dmartinol replied:
Good catch, fixing and revalidating it

mposolda commented on services/src/main/java/org/keycloak/broker/oidc/OIDCIdentityProvider.java:
```diff
@@ -389,6 +389,37 @@ public BrokeredIdentityContext getFederatedIdentity(String response) {
+                String filterName = getConfig().getClaimFilterName();
+                String filterValue = getConfig().getClaimFilterValue();
+
+                logger.infof("Filtering user %s by %s=%s", idToken.getOtherClaims().get(getusernameClaimNameForIdToken()), filterName, filterValue);
```
Is it possible to use `tracef` instead of `infof` (in this place and also in all other in this class) and also use `debugf` instead of `warnf` (in this place and also in all other in this class). As AFAIK `info` and `warn` are usually for special messages or server-wide messages/issues, but not for regular "user" requests.

dmartinol replied:
Thanks for sharing these practices

mposolda reviewed (changes requested):
I've added few additional inline comments in addition to @stianst comment.

I am maring this as "missing tests". We have identity brokering tests in the testsuite in the package `org.keycloak.testsuite.broker` . Hope you can use some existing tests for the inspiration.

pedroigor reviewed (commented):
This looks somewhat related to https://openid.net/specs/openid-connect-core-1_0.html#ClaimsParameter. 

However, the `claims` parameter is `OPTIONAL` and I'm not sure if OPs are supposed to *not* authenticate a user if a requested claim (and optionally its value) is not available for a given user.

I think it makes sense to at least make it more aligned with the idea of the `claims` parameter from the specification. Something like:

* Verify essential claim (ON/OFF. If enabled, ID tokens issued by the OP must have a specific claim. Otherwise, the user can not authenticate through this broker.)
* Essential claim (name)
* Essential claim value (value)

Even though not strictly related to `claims` support, using a similar terminology should help to understand what it does.

dmartinol commented on js/apps/admin-ui/public/locales/en/identity-providers-help.json:
```diff
@@ -38,6 +38,9 @@
   "trustEmail": "If enabled, email provided by this provider is not verified even if verification is enabled for the realm.",
   "accountLinkingOnly": "If true, users cannot log in through this provider.  They can only link to this provider.  This is useful if you don't want to allow login from the provider, but want to integrate with a provider",
   "hideOnLoginPage": "If hidden, login with this provider is possible only if requested explicitly, for example using the 'kc_idp_hint' parameter.",
+  "filteredByClaim": "If true, enables filter by claim, to reject authenticated users whose JWT claims do not include the configured claim filter",
```
will do, even if it might change if we also include Stian's comment above:
> Perhaps it could be renamed to something like Require claim to authenticate and Required claim name / Required claim value?

dmartinol commented on services/src/main/java/org/keycloak/broker/oidc/OIDCIdentityProvider.java:
```diff
@@ -389,6 +389,37 @@ public BrokeredIdentityContext getFederatedIdentity(String response) {
+
+                logger.infof("Filtering user %s by %s=%s", idToken.getOtherClaims().get(getusernameClaimNameForIdToken()), filterName, filterValue);
+                boolean hasKey = false;
+                for (String claim : idToken.getOtherClaims().keySet()) {
```
No special reasons. I will update the code and the documentation as well

dmartinol committed:
fix: integrated changes from PR

dmartinol committed:
Merge branch 'keycloak:main' into main

dmartinol commented:
> I am marking this as "missing tests". We have identity brokering tests in the testsuite in the package `org.keycloak.testsuite.broker` . Hope you can use some existing tests for the inspiration.

@mposolda while writing the test code I verified that `UserRepresentation.setAttributes` method has a `Map<String, List<String>>`, as if for each user attribute we could define multiple values, which I think it's not doable from the Keycloak Admin console. As a result, the OIDC provider receives an ID token claim that is a JSON array and cannot easily compare it with the filter configured in the OIDC configuration.
I can change the provider to look at the actual type of claim value, in this case a `java.util.List`, and apply the match on any of the contained values, but I'd also like to understand why I saw this behavior only with the test code while the debugging on the development server only returned claim values of `String` type.

dmartinol committed:
fix: added logic to manage token claim values of type List and integration test methods

dmartinol commented:
@pedroigor @mposolda integrated requested changed (apart from labels for which I'm waiting for some final decision)

dmartinol committed:
Merge branch 'main' into main

dmartinol committed:
Merge branch 'main' into main

dmartinol committed:
Merge branch 'main' into main

dmartinol committed:
Merge branch 'keycloak:main' into main

pedroigor commented:
@dmartinol I think the remaining tasks are basically around the names you are using to configure the essential|required claims.

I would use something like what I proposed (and it seems @mposolda agrees about it too). It is also aligned with the request from @stianst to make it more meaningful.

dmartinol committed:
Merge branch 'keycloak:main' into main

dmartinol committed:
Updated field names as per PR comments

dmartinol committed:
Merge branch 'main' into main

dmartinol commented:
> @dmartinol I think the remaining tasks are basically around the names you are using to configure the essential|required claims.
> 
> I would use something like what I proposed (and it seems @mposolda agrees about it too). It is also aligned with the request from @stianst to make it more meaningful.

@mposolda @pedroigor @stianst I believe I have submitted all required changes including integration testing, could you please clarify if anything else is missing or if there are other concerns to move forward with this PR?

jonkoops commented on js/apps/admin-ui/src/identity-providers/add/AdvancedSettings.tsx:
```diff
@@ -125,6 +136,92 @@ export const AdvancedSettings = ({ isOIDC, isSAML }: AdvancedSettingsProps) => {
+                labelOff={t("common:off")}
+                isChecked={field.value === 'true'}
+                onChange={(value) => {
+                    const v=""+value
+                    field.onChange(v)
```
Instead of declaring a new variable and concatenating a string to do a cast, you could call [`toString()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Boolean/toString) instead.
```suggestion
                    field.onChange(value.toString())
```

jonkoops commented on js/apps/admin-ui/src/identity-providers/add/AdvancedSettings.tsx:
```diff
@@ -125,6 +136,92 @@ export const AdvancedSettings = ({ isOIDC, isSAML }: AdvancedSettingsProps) => {
+                    setClaimFilterRequired(value)
+                  }
+                }
+                aria-label="filteredByClaim"
```
An `aria-label` attribute should have a human readable value.

jonkoops commented on js/apps/admin-ui/src/identity-providers/add/AdvancedSettings.tsx:
```diff
@@ -125,6 +136,92 @@ export const AdvancedSettings = ({ isOIDC, isSAML }: AdvancedSettingsProps) => {
+                ? ValidatedOptions.error
+                : ValidatedOptions.default
+              }
+              {...register("config.claimFilterValue", { required: claimFilterRequired })}
```
I am not sure you can update the `required` value here after it has been passed. It might have to be set explicitly with a different API.

Have you tested that this works as expected?

jonkoops commented:
@dmartinol there are several linting errors in the front-end code, could you fix these?

dmartinol commented:
> @dmartinol there are several linting errors in the front-end code, could you fix these?

sure, I will give it a try, thanks for reviewing the changes

dmartinol committed:
Merge branch 'keycloak:main' into main

dmartinol committed:
Update js/apps/admin-ui/src/identity-providers/add/AdvancedSettings.tsx

Co-authored-by: Jon Koops <jonkoops@gmail.com>

dmartinol committed:
Update js/apps/admin-ui/src/identity-providers/add/AdvancedSettings.tsx

Co-authored-by: Jon Koops <jonkoops@gmail.com>

dmartinol committed:
Integrated recommended changes

dmartinol commented:
@jonkoops where can I see the `linting errors in the front-end code`?
Submitted the other changes in the meantime

dmartinol committed:
linting js

dmartinol committed:
Merge branch 'keycloak:main' into main

jonkoops commented on js/apps/admin-ui/src/identity-providers/add/AdvancedSettings.tsx:
```diff
@@ -125,6 +140,89 @@ export const AdvancedSettings = ({ isOIDC, isSAML }: AdvancedSettingsProps) => {
+                onChange={(value) => {
+                  field.onChange(value.toString());
+                }}
+                aria-label="filteredByClaim"
```
This can be removed as the field is already labeled through `FormGroupField`.

dmartinol replied:
Removing "aria-label" field thanks

jonkoops commented on js/apps/admin-ui/src/identity-providers/add/AdvancedSettings.tsx:
```diff
@@ -93,8 +98,18 @@ type AdvancedSettingsProps = { isOIDC: boolean; isSAML: boolean };
+  const filteredByClaim = useWatch({
+    control,
+    name: "config.filteredByClaim",
+    defaultValue: false,
```
Not sure, but likely this should be a string to match the actual value stored in the form.
```suggestion
    defaultValue: "false",
```

dmartinol replied:
Not sure why it worked before but you are correct, it's a string

dmartinol committed:
Update js/apps/admin-ui/src/identity-providers/add/AdvancedSettings.tsx

Co-authored-by: Jon Koops <jonkoops@gmail.com>

dmartinol committed:
Removed redundant aria-label field

dmartinol committed:
Merge branch 'main' into main

jonkoops reviewed (dismissed):
Changes from the front-end side of things are looking good to me.

dmartinol commented:
@mposolda looks like you requested one more change but, apart from the test code that was already submitted, I can't see any other requests, could you please verify if I'm missing anything?

dmartinol committed:
Merge branch 'main' into main

dmartinol committed:
Merge branch 'main' into main

mposolda commented:
@dmartinol There are failing tests and looks that they are regression of this PR. I suggest to rebase and fix the failing tests. IMO this PR is good to go once that is done.

dmartinol committed:
Merge branch 'main' into main

dmartinol committed:
Merge pull request #1 from keycloak/main

Simplify TokenManager expiration calculations using SessionExpiration…

dmartinol committed:
Merge branch 'main' into main

dmartinol committed:
Fixed broken tests

dmartinol committed:
Merge branch 'main' into main

dmartinol commented:
> Details

@mposolda now seems that some "Admin UI E2E" tests are failing, do I need to install cypress and fix the broken test `identity_providers_test.spec.ts`?

jonkoops commented:
@dmartinol yes, if possible please also fix the tests. You can find instructions on how to run the tests in [the README](https://github.com/keycloak/keycloak/tree/main/js/apps/admin-ui#integration-testing-with-cypress).

dmartinol committed:
Merge branch 'main' into main

dmartinol committed:
Merge branch 'main' into main

dmartinol committed:
Fixed UI test for IDP

dmartinol commented:
> @dmartinol There are failing tests and looks that they are regression of this PR. I suggest to rebase and fix the failing tests. IMO this PR is good to go once that is done.

Fixed broken tests: the page became longer and the initial switched in the advanced settings page were not visible, so I added a click on the `Advanced Settings` link to restore the page.
Also added check that the new switch and text fields are populated

jonkoops reviewed (changes requested):
There are some linting issues in the front-end that need to be resolved.

dmartinol committed:
Fixed linting issues

dmartinol committed:
fixing lint

dmartinol committed:
linting fixes

dmartinol committed:
linting

dmartinol committed:
added extra call to ensureAdvancedSettingsAreVisible()

dmartinol committed:
fixed should revert and save options test

dmartinol commented:
> There are some linting issues in the front-end that need to be resolved.

It took me 38 commits, but fInally "All checks have passed"

trixpan commented:
why not using "Post login flow" for achieving the same goal?

trixpan commented:
But isn't the use creation controlled by First Login Flow? 

That seems like a better fit than creating an option that will be tightly linked to the IdP section of the login flow.

jonkoops reviewed (dismissed):
Changed to the front-end LGTM

andymunro reviewed (approved):
Looks good, @jonkoops

trixpan commented:
> you mean to develop a new Condition to validate the required claim and put it before the "Create User if Unique" to exit the flow in case the claim is not matched?

`Create User if Unique` is not a requirement of that flow, many deployments will simply link to existing users using `Detect Existing Broker User` followed by auto linking as documented. 

I would say that a conditional, added to the flow, would be far more natural than a filter linked to a particular IdP.

Haven't tested it yet, but I wouldn't be surprised `Condition - User Attribute` execution would work?

bmozaffa commented:
> Haven't tested it yet, but I wouldn't be surprised `Condition - User Attribute` execution would work?

Definitely needs to be verified, but at least my impression was that the user is created locally before IdP claims are imported as user attributes, so this would be a chicken and egg problem if we're trying to avoid initial login/user creation

dmartinol commented:
> Haven't tested it yet, but I wouldn't be surprised `Condition - User Attribute` execution would work?

If we use that condition it means that a) we created an IDP mapper to convert the claim to the attribute and b) (I guess) the user has to be created together with attributes
So this would not solve the request to not create the user

trixpan commented:
> > Haven't tested it yet, but I wouldn't be surprised `Condition - User Attribute` execution would work?
> 
> If we use that condition it means that a) we created an IDP mapper to convert the claim to the attribute and b) (I guess) the user has to be created together with attributes So this would not solve the request to not create the user

If that is the case, again, haven't tested, you don't need to use that particular step provider. Aby other purpose built execution provider would be perfectly fine. 

The point I am trying to make is: there is a flow for post authentication, that flow clearly has steps that happen before the user is created (as you may chose to use the existing user create a new one). Sticking to that pattern of flow based checks seems cleaner that bolting something into the IdP implementation.

In the end, the result is the same. One (via flow) will work will all providers, the other one will depend on class inheritance or re-implementation. But access will be conditional to attributes.

pedroigor commented:
@trixpan I share the same but we can't guarantee that all protocols can share the same mechanism for handling user attributes. It should work if we enable it in the `IdpCreateUserIfUniqueAuthenticator` authenticator, for both OIDC and SAML. There the user is still transient and we can access the claims/attributes.

For now, I would keep it specific to OIDC and enable support for SAML if there are demands for it.

dmartinol commented:
@mposolda seems that there are more changes required from you but cannot find anything. could you please verify and let me know what else is missing?

stianst commented:
Would be worth adding this to release notes (https://github.com/keycloak/keycloak/blob/main/docs/documentation/release_notes/topics/22_0_0.adoc), as I know this has been raised a few times at least in the past.

mposolda commented:
@dmartinol LGTM, Thanks for the tests! As long as it is added to release notes as proposed by @stianst , we should be good with this PR from my point of view.

@pedroigor @trixpan @dmartinol Handling this at the first-broker-login-flow/post-broker-login-flow may be also possible, but I am not sure if it's better than adding the config options directly into the Idp as done in this PR. Both approaches have advantages and disadvantages though... IMO flow approach might be conceptually cleaner and is independent on protocol, however it can result in many flows created (in case you have many Identity providers and every of them has slightly different set of "required claims").

dmartinol committed:
Merge branch 'keycloak:main' into main

dmartinol committed:
Updated release notes

mposolda reviewed (dismissed):
@dmartinol Thanks! LGTM. As long as tests are OK, I will need to squash the commits...

dmartinol committed:
Merge branch 'main' into main

dmartinol committed:
Merge branch 'main' into main

andymunro commented on docs/documentation/release_notes/topics/22_0_0.adoc:
```diff
@@ -127,11 +127,19 @@ q=<name>:<value> <name>:<value> ...
+OpenID Connect identity providers support a new configuration to specify that the ID tokens issued by the identity provider must have a specific claim,
+otherwise the user can not authenticate through this broker.
+
+The option is disabled by default, when it is enabled we can specify the name of the JWT token claim to filter and the value to match 
```
```suggestion
The option is disabled by default; when it is enabled, you can specify the name of the JWT token claim to filter and the value to match 
```

andymunro commented on docs/documentation/server_admin/topics/identity-broker/configuration.adoc:
```diff
@@ -58,6 +58,16 @@ Although each type of identity provider has its configuration options, all share
 
+
+|Verify essential claim
+|When *ON*, ID tokens issued by the identity provider must have a specific claim. Otherwise, the user can not authenticate through this broker.
+
+|Essential claim
+|When *Verify essential claim* is *ON*, the name of the JWT token claim to filter (match is case sensitive).
+
+|Essential claim value
+|When *Verify essential claim* is *ON*, the value of the JWT token claim to match (supports regular expression format).
```
```suggestion
|When *ON*, ID tokens issued by the identity provider must have a specific claim. Otherwise, the user can not authenticate through this broker

|Essential claim
|When *Verify essential claim* is *ON*, the name of the JWT token claim to filter (match is case sensitive)

|Essential claim value
|When *Verify essential claim* is *ON*, the value of the JWT token claim to match (supports regular expression format)
```

andymunro reviewed (changes requested):
Looks good, @jonkoops. A few minor changes to suggest.

dmartinol committed:
Update docs/documentation/server_admin/topics/identity-broker/configuration.adoc

Co-authored-by: andymunro <48995441+andymunro@users.noreply.github.com>

dmartinol committed:
Update docs/documentation/release_notes/topics/22_0_0.adoc

Co-authored-by: andymunro <48995441+andymunro@users.noreply.github.com>

dmartinol committed:
Merge branch 'main' into main

dmartinol committed:
Merge branch 'main' into main

dmartinol commented:
> @jonkoops @dmartionol Thanks for the updates, but I am not sure if test failures are just "instability" or if it's regression of this PR? Seems to me that issues in `Admin UI E2E (3, Chrome)` are regression of this PR (Seeing 7 failures in `identity_providers_test.spec.ts` in this group). There is also something in `Admin UI E2E (1, Chrome)`, but looks only as single test failure, so might be instability? Could you confirm?

Now `users_test.spec.ts` is failiing, and I don't think this is affected by the PR, is it? Anyway, I also see random failures on my local env, so I'm not sure what's going on.

dmartinol committed:
Merge branch 'main' into main

mposolda commented:
@dmartinol I think that those test failures are not caused by your PR, but rather some general instability as I see them also in some other PR https://github.com/keycloak/keycloak/pull/20621 (looks like exactly same test failures)... @jonkoops Could you please confirm?

BTV. It can be helpful if you stop merging this PR with latest Keycloak main (without any changes in your PR) as it just re-triggers the build when the links to the old build would be lost. We can eventually re-trigger your jobs when needed and/or will ask you for rebase/merge.

mposolda committed:
Merge branch 'main' into main

mposolda commented:
@dmartinol I've updated the branch to latest Keycloak main. That should probably help to fix test failures (They were fixed in the meantime in Keycloak main by PR https://github.com/keycloak/keycloak/pull/21312 ). Hopefully we're finally very close to merge now :-)

mposolda committed:
Merge branch 'main' into main

mposolda committed:
Merge branch 'main' into main
